Cubical/Algebra/ZariskiLattice/Properties.agda

```diff
@@ -0,0 +1,221 @@
+{-# OPTIONS --safe --lossy-unification #-}
+module Cubical.Algebra.ZariskiLattice.Properties where
+
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.HLevels
+open import Cubical.Foundations.Powerset using (ℙ)
+                                         renaming (_∈_ to _∈ₚ_ ; ∈-isProp to ∈ₚ-isProp)
+
+open import Cubical.Data.Nat using (ℕ)
+open import Cubical.Data.Sigma.Properties
+open import Cubical.Data.FinData
+open import Cubical.Relation.Binary.Order.Poset
+
+open import Cubical.Algebra.CommRing
+open import Cubical.Algebra.CommRing.Localisation
+open import Cubical.Algebra.CommRing.Ideal
+open import Cubical.Algebra.CommRing.FGIdeal
+open import Cubical.Algebra.CommRing.RadicalIdeal
+open import Cubical.Tactics.CommRingSolver
+open import Cubical.Algebra.Semilattice
+open import Cubical.Algebra.Lattice
+open import Cubical.Algebra.DistLattice
+open import Cubical.Algebra.DistLattice.Downset
+
+open import Cubical.Algebra.ZariskiLattice.Base
+open import Cubical.Algebra.ZariskiLattice.UniversalProperty
+
+open import Cubical.HITs.SetQuotients as SQ
+import Cubical.HITs.PropositionalTruncation as PT
+
+
+private variable ℓ : Level
+
+module _ (R : CommRing ℓ) where
+  open Iso
+  open CommRingStr ⦃...⦄
+  open PosetStr ⦃...⦄
+
+  open Exponentiation R
+  open CommIdeal R
+  open RadicalIdeal R
+
+  open ZarLat R
+  open ZarLatUniversalProp R
+
+  open JoinSemilattice (Lattice→JoinSemilattice (DistLattice→Lattice ZariskiLattice))
+       using (IndPoset)
+
+  private
+    instance
+      _ = R .snd
+      _ = IndPoset .snd
+
+    ⟨_⟩ : {n : ℕ} → FinVec (fst R) n → CommIdeal
+    ⟨ V ⟩ = ⟨ V ⟩[ R ]
+
+  unitLemmaZarLat : ∀ f → D f ≡ D 1r → f ∈ₚ R ˣ
+  unitLemmaZarLat f Df≡D1 = PT.rec (∈ₚ-isProp (R ˣ) f) radicalHelper 1∈√⟨f⟩
+    where
+    D1≤Df : D 1r ≤ D f
+    D1≤Df = subst (_≤ D f) Df≡D1 (is-refl _)
+
+    1∈√⟨f⟩ : 1r ∈ √ ⟨ replicateFinVec 1 f ⟩
+    1∈√⟨f⟩ = isEquivRel→effectiveIso ∼PropValued ∼EquivRel _ _ .fun D1≤Df .fst zero
+
+    radicalHelper : Σ[ n ∈  ℕ ] 1r ^ n ∈ ⟨ replicateFinVec 1 f ⟩ → f ∈ₚ R ˣ
+    radicalHelper (n , 1ⁿ∈⟨f⟩) = PT.rec (∈ₚ-isProp (R ˣ) f) fgHelper 1ⁿ∈⟨f⟩
+      where
+      fgHelper : Σ[ α ∈ FinVec (fst R) 1 ] 1r ^ n ≡ linearCombination R α (replicateFinVec 1 f)
+               → f ∈ₚ R ˣ
+      fgHelper (α , 1ⁿ≡α₀f+0) = α zero , path
+        where
+        path : f · α zero ≡ 1r
+        path = f · α zero      ≡⟨ solve! R ⟩
+               α zero · f + 0r ≡⟨ sym 1ⁿ≡α₀f+0 ⟩
+               1r ^ n          ≡⟨ 1ⁿ≡1 n ⟩
+               1r ∎
```

Comment:
About half of this proof is proving that if `1 ∈ √⟨f⟩` then `f` is a unit. It would be good to have that in a reusable form. Perhaps as two lemmas (in appropriate files): (1) if `1 ∈ √I` then `1 ∈ I`, (2) if `1 ∈ ⟨f⟩` then `f` is a unit.

Reply:
Yeah agreed, I'll have a look at it later.

Reply:
Is something like this what you had in mind :
https://github.com/agda/cubical/pull/1096/commits/ddc4fa53198b8eb3c3bd8b0cb55c1680648d3e93

Reply:
Yes, very nice!